src/mtype.c

```diff
@@ -6872,10 +6872,12 @@ Type *TypeTypeof::syntaxCopy()
 Dsymbol *TypeTypeof::toDsymbol(Scope *sc)
 {
     //printf("TypeTypeof::toDsymbol('%s')\n", toChars());
-    Type *t = semantic(loc, sc);
-    if (t == this)
-        return NULL;
-    return t->toDsymbol(sc);
+    Expression *e;
+    Type *t;
+    Dsymbol *s;
+    resolve(loc, sc, &e, &t, &s);
+
+    return s;
```

Comment:
Looks good. 

I noticed that same bug exists in `TypeReturn::toDsymbol()`. Could you also fix that?

```d
class C14348
{
    int foo() { return 1; }
}

C14348 foo14348()
{
    alias foo = typeof(return).foo;  // currently doesn't work.
    return null;
}
```